File: 0135 Candy/solution.py
#! python3

class Solution(object):
    def candy(self, ratings):
        """
        :type ratings: List[int]
        :rtype: int
        """
        res = [1] * len(ratings)
        
        for i in range(1, len(ratings)):
            if ratings[i-1] < ratings[i]:
                res[i] = res[i-1] + 1
        
        for i in range(len(ratings)-1, 0, -1):
            if ratings[i-1] > ratings[i] and res[i-1] <= res[i]:
                res[i-1] = res[i] + 1
        
        return sum(res)

if __name__ == "__main__":
    print(Solution().candy([1, 0, 2]))

File: 0135 Candy/solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int candy(vector<int>& ratings) {
        int n = ratings.size();
        vector<int> f(n, 1);

        for (int i = 1; i < n; i ++)
            if (ratings[i - 1] < ratings[i])
                f[i] = f[i - 1] + 1;
        for (int i = n - 2; i >= 0; i --)
            if (ratings[i] > ratings[i + 1] && f[i] <= f[i + 1])
                f[i] = f[i + 1] + 1;

        int res = 0;
        for (int n : f) res += n;
        return res;
    }
};

int main(){
    vector<int> ratings{1, 0, 2};
    cout << Solution().candy(ratings) << endl;
    return 0;
}